### ISSUE TITLE
Improving the existing bloom filter without involving refactoring

### ISSUE THREAD

**Issue by KaguraMilet** [slatedb/slatedb#229]
**Is your feature request related to a problem? Please describe.**
This issue involves some optimization of the current filter implementation without refactoring it.
#### Reducing potential function call overhead
In the current Bloom filter implementation, `BloomFilterBuilder` holds the `key_hashes` field. In the SST build process this requires us to call `self.filter.add_key(key)` every time. Compiler may be able to inline this function,  but we cannot guarantee that. As a result, this can introduce potential function call overheads.
[SST add function](https://github.com/slatedb/slatedb/blob/main/src/sst.rs#L405)

#### Use better hash function
In [#223](https://github.com/slatedb/slatedb/pull/223#issuecomment-2365061314), I mentioned that the current bloom filter uses sip hash as the default hash function. [Sip Hash in slateDB](https://github.com/slatedb/slatedb/blob/main/src/filter.rs#L85)
But according to smhasher tests, siphash doesn't really perform well compared to xxhash, farmhash, murmur, etc. According to the results of Boston's paper, the false positive rate of their choice of murmur64 as the hash function is extremely close to the theoretical optimum.
![hash_filter](https://github.com/user-attachments/assets/bd386c38-8ce2-4b04-971d-41cabdc32b26)
RocksDB uses xxhash as default hash function [RocksDB hash.cc](https://github.com/facebook/rocksdb/blob/main/util/hash.cc#L85)

**Describe the solution you'd like**
#### Reducing potential function call overhead
For the first problem, we can let `EncodedSsTableBuilder` holds the `key_hashes` field. Then when we build the SST we just pass the `key_hashes` field directly into BloomFilterBuilder::build. So we don't need to call `BloomFilterBuilder::add_key` frequently.
[SlateDB bloom filter build](https://github.com/slatedb/slatedb/blob/main/src/filter.rs#L37)

#### Replace current hash function
We can perform relevant benchmarks such as xxhash, murmur, farmhash. But unfortunately I don't have the time to complete the benchmarks at the moment, so maybe this is a good opportunity to encourage people to make their first contribution to slateDB.

**Describe alternatives you've considered**
None

**Additional context**
None

**Comment by criccomini** [comment]
SGTM. Let's hold off on this until we have some basic perf testing/benchmarking set up. I just opened #230 to get a basic `cargo bench` command. In addition, @taburet is working on an S3 benchmark similar to our compaction bench command.

**Comment by criccomini** [comment]
We probably want a `cargo bench` benchmark for this one. I have some stub code in this abandoned PR:

https://github.com/slatedb/slatedb/pull/232

The test in that PR is for puts/gets, which I don't think is useful. But a `cargo bench` test for the bloom filter would be good.

**Comment by criccomini** [comment]
Talked with @AlvinKuruvilla. Sounds like he's going to take this one.

**Comment by AlvinKuruvilla** [comment]
I appreciate it @criccomini. I'll take a stab at this when I get some time to work on it and circle back if I have any updates or questions :)

**Comment by criccomini** [comment]
Sounds good! Feel free to break the work up (like, just add the benchmark as one PR, then dealing with the hash comparisons separately).

**Comment by KaguraMilet** [comment]
> I appreciate it @criccomini. I'll take a stab at this when I get some time to work on it and circle back if I have any updates or questions :)

Maybe you don't need a very complicated end-to-end test, I think you can separate the filter module and execute a microbench by using [flamegraph](https://github.com/flamegraph-rs/flamegraph), [samply](https://github.com/mstange/samply/), criterion and son on.

For the first idea, I have an implementation, and I use split block bloom filter inspired by Apache Parquet instead of standard bloom filter. @AlvinKuruvilla 
[Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=74b232805018d7a30f21b1b590f2c39c)

**Comment by criccomini** [comment]
@KaguraMilet, yes I pointed to my abandoned PR #232, which uses Criterion. I think that should be fine.

**Comment by AlvinKuruvilla** [comment]
#250 is my first crack at the issue. I think the attached flamegraph, illustrates (at least at a high level) the potential performance issues of our current hash function. But I'm totally open to suggestions and feedback :)
